### PR TITLE
CA-280242: Open destination VDI O_DIRECT (as well as source).

### PR DESCRIPTION
**Pull request by jonludlam**
Without O_DIRECT we weren't getting IO errors propagated from the
sendfile(2) call (not even on close). This was causing us not to notice
when copying VDIs to full thinly-provisioned SRs.

Signed-off-by: Jon Ludlam <jonathan.ludlam@citrix.com>

### CODE REVIEW

File: src/impl.ml [hunk 1]
```diff
@@ -673,7 +673,8 @@ let write_stream common s destination source_protocol destination_protocol preze
   let use_ssl = match endpoint with Https _ -> true | _ -> false in
   ( match endpoint with
     | File path ->
-      Lwt_unix.openfile path [ Unix.O_RDWR; Unix.O_CREAT ] 0o0644 >>= fun fd ->
+      let unix_fd = Vhd_format_lwt.File.openfile path true 0o644 in
```
Comment (edwintorok) [inline src/impl.ml]: What happens to O_CREAT here, and is it relevant?

Comment (jonludlam) [reply src/impl.ml]: In our case the file has already been created. It's possible we might want to follow the pattern later in the file: https://github.com/jonludlam/vhd-tool/blob/f039b64a025fd546519cc434bc6033d38509b54b/src/impl.ml#L968-L972

Comment (edwintorok) [reply src/impl.ml]: Sounds good, factor out that out into an `open_odirect_rw` function?

**Comment by jonludlam** [comment]
@gaborigloi - `use_unbuffered` is definitely set by the time we use it: we're using the same function to open the source VDI, and that was already opened O_DIRECT before this change.

File: src/impl.ml (hunk 1 above)
Comment (jonludlam) [reply src/impl.ml]: I can't quite use the same function - that does `Vhd_format_lwt.IO.openfile` rather than a `Vhd_format_lwt.File.openfile` - the former returns an opaque type encapsulating the fd that I don't think I can break apart.

Comment (jonludlam) [reply src/impl.ml]: Oh!, my mistake, there it is:

https://github.com/xapi-project/ocaml-vhd/blob/master/vhd_format_lwt/iO.mli#L22

Comment (lindig) [reply src/impl.ml]: From the diff it is not obvious that `O_DIRECT` is now being used as the commit message suggests. I think this needs to explained either in the commit message or in a comment.

Comment (jonludlam) [reply src/impl.ml]: it's awkward because the o_directness is dictated by a global var, so it's be a bit of a misnomer. I've named it 'open_with_creat' instead.

@lindig I'll update the commit message when I squash these after everyone is happy with the code.

File: src/impl.ml [hunk 2]
```diff
@@ -141,6 +141,14 @@ let create common filename size parent =
   with Failure x ->
     `Error(true, x)
 
+let open_with_creat path rw =
+  ( if not(Sys.file_exists path) then begin
```
Comment (edwintorok) [inline src/impl.ml]: I don't like mixing blocking functions from `Sys` with non-blocking functions from Lwt. This should use `Lwt_unix.file_exists` instead, however that was a pre-existing problem and should be done in a separate PR.

Comment (jonludlam) [reply src/impl.ml]: agreed.

**Comment by jonludlam** [comment]
OK, alll squashed and commit msg updated.

File: src/impl.ml (hunk 2 above)
Comment (lindig) [reply src/impl.ml]: Is it wise to rely on a global variable in another library to get the desired flags? I would have liked it better to do this explicitly here, if it is so important.
